[Transform] Auto retry Transform start

prwhelan opened the pull request:
Currently, Transforms can fail to start due to failure to load the Config from its internal index.  This usually happens when a Transform is created and immediately started by a system.  The error looks like:

```
Failed to load transform configuration for transform [id]
```

Now, we will automatically retry the startup logic until the Config is ready.

Some notes:
- We cannot determine if a transform is unattended or not, so at this stage we will assume all transforms are unattended.
- The persistent task running the transform will move into the `STARTED` state.  Users can stop the persistent task and retry logic using the Transform's Stop API.
- Transforms can still only be started once.  A subsequent call to start the Transform, even while it is retrying, will still fail.
- While retrying, the Transform will report `Yellow` health in the API and `degraded` in Kibana.  The health message will include that the transform is automatically retrying and what error it had encountered.
- The Transform's audit log will include an info statement that a retry was scheduled and include what error it had encountered.
- The first retry will happen before 1 minute has passed, and then at a random time up until two minutes after that for ~292 years. The random retry interval should help stagger the scheduled actions to keep the system healthy.

----------------------------

Testing:

Calling the stats API
```
"health": {
  "status": "yellow",
  "issues": [
    {
      "type": "transform_starting",
      "issue": "Transform task is automatically retrying its startup process",
      "details": "Retrying Transform action due to error: java.lang.IllegalStateException: Failed to load transform configuration for transform [last-log-from-clientip].",
      "count": 1,
      "first_occurrence": 1710250715674
    }
  ]
}
```

Kibana:
![retryunattended](https://github.com/elastic/elasticsearch/assets/1223797/14eb3007-bffa-4ec9-a4b9-5145e0d2862b)

prwhelan committed:
[Transforms] Auto retry unattended Transform start

Currently, unattended Transforms can fail to start due to failure to
load the Config from its internal index.  This usually happens when a
Transform is created and immediately started by a system.  The error
looks like:

```
Failed to load transform configuration for transform [id]
```

Now, we will automatically retry the startup logic until the Config is
ready.

Some notes:
- We cannot determine if a transform is unattended or not, so at this
  stage we will assume all transforms are unattended.
- The persistent task running the transform will move into the `STARTED`
  state.  Users can stop the persistent task and retry logic using the
  Transform's Stop API.
- While retrying, the Transform will report `Yellow` health in the API
  and `degraded` in Kibana.  The health message will include that the
  transform is automatically retrying and what error it had encountered.
- The Transform's audit log will include an info statement that a retry
  was scheduled and include what error it had encountered.
- The first retry will happen before 1 minute has passed, and
  then at a random time up until two minutes after that for ~292 years.
  The random retry interval should help stagger the scheduled actions to
  keep the system healthy.

prwhelan committed:
Update docs/changelog/106243.yaml

elasticsearchmachine commented:
Hi @prwhelan, I've created a changelog YAML for you.

prwhelan committed:
Remove Machine Learning from changelog area

prwhelan committed:
Clean up tests a bit

prwhelan committed:
Fix checkstyle

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

prwhelan commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformRetryableActions.java:
```diff
@@ -0,0 +1,204 @@
+ * This state is not expected to be persisted across the cluster or persisted across cluster restarts.
+ * This class is thread safe.
+ */
+public class TransformRetryableActions {
```
Yeah, we could, I picked Map + `RetryableAction` instead of `TransformScheduler` mostly because I didn't want to modify `TransformScheduler` to have the new needed functionality:
1. The ability for a `TransformScheduledTask` to mark itself "finished" and remove itself from the queue, which we could do by having the triggered listener return a boolean to say if it should keep running.
2. The random interval between retries, which would be fairly easy to write since we always calculate the next execution time.
3. Some way for users to visibly see that a retry is happening while the persistent task is started but the transform indexer is not, which would be some new API in the scheduler.

RetryableAction already had the pattern for 1 and 2, so it just needed 3 and the global storage (Map).

Now that we have a working version, I could maybe timebox another iteration that uses `TransformScheduler` with the above changes.  Having fewer moving parts would be good.

prwhelan replied:
This was kinda my result of this: https://github.com/prwhelan/elasticsearch/commit/4ea9842e404d132e3ca73708ce63d83a9771a062

There's no 2 or 3 yet, but that's about what it would look like.  There may be some more adjustments within the Scheduler to not log/audit, otherwise every transform start would see audits for register(startup) -> deregister(startup) -> register(start)

przemekwitek replied:
> Now that we have a working version, I could maybe timebox another iteration that uses TransformScheduler with the above changes. Having fewer moving parts would be good.

I agree. Since we already have `TransformScheduler` class (and, BTW, we own it so we can extend it with any functionality that makes sense, e.g.: add random jitter to scheduled time), it would be good to re-use it not to introduce yet another retrying mechanism.

prwhelan commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformRetryableActions.java:
```diff
@@ -0,0 +1,204 @@
+     */
+    void stopAll(String transformId) {
+        // copy to an temp list, avoiding a modification exception since the cancel function will run a listener that removes from the map
+        retriesForTransform(transformId).toList().forEach(task -> {
```
Yes, because `cancel` will eventually call the above listener that removes the entry from the map, so we'd be modifying the map during the iteration.  We need the extra List so that we can iterate over the tasks while updating the map.

prwhelan commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformPersistentTasksExecutor.java:
```diff
@@ -392,6 +396,32 @@ private static void markAsFailed(TransformTask task, Throwable exception, String
+                    newState,
+                    ActionListener.wrap(
+                        rr -> logger.debug("[{}] marked as retrying in TransformState.", task.getTransformId()),
+                        ee -> logger.warn(() -> format("[%s] failed to persist state.", task.getTransformId()), ee)
```
It's to get the throwable in there as well, but as I learned from your PR yesterday, I could use `logger.atWarn().withThrowable(ee).log("[{}] failed to persist state.", task.getTransformId())`

prwhelan committed:
Use logbuilder; requireNonNull

prwhelan committed:
Use TransformScheduler for retries

  Rather than maintain a map and use RetryActions, we are using the
  TransformScheduler as the retry mechanism.  TransformScheduler is
  responsible for periodically retrying the startup task based on the
  Transform's frequency.

  Startup failures are now reported out through the Context object.

  Removed most of the logic around creating retry functions.  We may never
  need this, but we can always add it back if we encounter a new startup
  function to retry.

przemekwitek commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformHealthChecker.java:
```diff
@@ -38,7 +38,8 @@ public enum IssueType {
         TRANSFORM_INDEXER_FAILED("Transform indexer failed"),
-        TRANSFORM_INTERNAL_STATE_UPDATE_FAILED("Task encountered failures updating internal state");
+        TRANSFORM_INTERNAL_STATE_UPDATE_FAILED("Task encountered failures updating internal state"),
+        TRANSFORM_STARTING("Transform task is automatically retrying its startup process");
```
For consistency, could you rename this constant to `TRANSFORM_STARTUP_FAILED` or similar?

przemekwitek commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformHealthChecker.java:
```diff
@@ -145,6 +145,24 @@ public static TransformHealth checkTransform(TransformTask transformTask, @Nulla
             );
         }
 
+        if (transformContext.getStartUpFailureCount() != 0) {
```
Could you add a unit test for this case to `TransformHealthCheckerTests`?

przemekwitek commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformContext.java:
```diff
@@ -214,6 +217,37 @@ Instant getLastStatePersistenceFailureStartTime() {
+        startUpFailureTime = null;
+    }
+
+    int getStartUpFailureCount() {
```
Could you add a unit test to `TransformContextTests` similar to `testFailureCount`?

przemekwitek reviewed (commented):
Could you add an integration test for the case when there are failures fetching transform config?
You can create one under `x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration`

prwhelan committed:
Addressing comments

prwhelan commented:
> Could you add an integration test for the case when there are failures fetching transform config? You can create one under `x-pack/plugin/transform/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration`

Is there a good way to simulate failures fetching transform config in the integration tests? I could not figure it out. I looked through most of the existing integration tests, but I couldn't find something that would effectively return an error in between when the `start` API is called and when the config is retrieved on the assigned node.

I added unit test that simulate a failure (along with a happy case).

przemekwitek reviewed (approved):
LGTM
